File: questions/rate-limiter/src/java/com/lld/ratelimit/SlidingWindowLog.java
package com.lld.ratelimit;

import java.util.ArrayDeque;

/**
 * One timestamp per allowed request, oldest at the front.
 *
 * ArrayDeque is the structure because requests only expire from the old end.
 * peekFirst is the oldest, removeFirst drops it, addLast records a new hit.
 * All three are O(1). A HashSet cannot find "older than T" without a scan.
 * A HashMap keyed only by timestamp collapses two requests in the same
 * millisecond into one entry. A TreeMap pays for ordered delete of an
 * arbitrary key, and this algorithm never deletes from the middle.
 *
 * The same shape in Redis is a sorted set: score is the timestamp, member
 * is a unique id. See redis/sliding-window-log.lua. A Redis list is the
 * wrong shared structure because it cannot delete by score in one command.
 */
public final class SlidingWindowLog implements Limiter {
    private final ArrayDeque<Long> hits = new ArrayDeque<>();
    private final long windowMs;
    private final int limit;

    public SlidingWindowLog(long windowMs, int limit) {
        this.windowMs = windowMs;
        this.limit = limit;
    }

    @Override
    public synchronized Decision tryAcquire(long nowMs) {
        long cutoff = nowMs - windowMs;
        while (!hits.isEmpty() && hits.peekFirst() <= cutoff) {
            hits.removeFirst();
        }
        if (hits.size() >= limit) {
            long retryAt = hits.peekFirst() + windowMs;
            return Decision.deny(limit, 0, retryAt - nowMs);
        }
        hits.addLast(nowMs);
        return Decision.allow(limit, limit - hits.size());
    }
}

File: questions/rate-limiter/src/redis/sliding-window-log.lua
-- One identity. One sorted set.
-- Score is the request time in milliseconds. Member is a unique id, never
-- the timestamp alone: two requests in the same millisecond would overwrite
-- one ZSET member and the limit would under-count.
--
-- KEYS[1]  the sorted set
-- ARGV[1]  now, ms
-- ARGV[2]  window, ms
-- ARGV[3]  limit
-- ARGV[4]  unique member
--
-- Returns {allowed, count, retryAfterMs}

local now = tonumber(ARGV[1])
local window = tonumber(ARGV[2])
local limit = tonumber(ARGV[3])
local member = ARGV[4]
local cutoff = now - window

redis.call('ZREMRANGEBYSCORE', KEYS[1], '-inf', cutoff)
local count = redis.call('ZCARD', KEYS[1])

if count >= limit then
  local oldest = redis.call('ZRANGE', KEYS[1], 0, 0, 'WITHSCORES')
  local retry = 0
  if oldest[2] ~= nil then
    retry = tonumber(oldest[2]) + window - now
    if retry < 0 then
      retry = 0
    end
  end
  redis.call('PEXPIRE', KEYS[1], window)
  return {0, count, retry}
end

redis.call('ZADD', KEYS[1], now, member)
redis.call('PEXPIRE', KEYS[1], window)
return {1, count + 1, 0}
